File: bench/benchmarkgen.c
/*
 * Copyright (c) 2011 Joseph Gaeddert
 * Copyright (c) 2011 Virginia Polytechnic Institute & State University
 *
 * This file is part of liquid.
 *
 * liquid is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * liquid is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with liquid.  If not, see <http://www.gnu.org/licenses/>.
 */

// 
// bench/benchmarkgen.c
// 
// Benchmark generator object definition
//

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

#include "benchmarkgen.h"

#define NAME_LEN            (256)
#define BENCHMARK_VERSION   "0.3.0"

struct benchmark_s {
    char name[NAME_LEN];
    struct package_s * package;
};

struct package_s {
    char name[NAME_LEN];
    struct benchmark_s * benchmarks;
    unsigned int num_benchmarks;
};

struct benchmarkgen_s {
    unsigned int num_packages;
    struct package_s * packages;
};

// create benchmark generator object
benchmarkgen benchmarkgen_create()
{
    benchmarkgen q = (benchmarkgen) malloc(sizeof(struct benchmarkgen_s));

    q->num_packages = 0;
    q->packages = NULL;

    // add defulat null package/benchmark
    benchmarkgen_addpackage(q, "null");
    benchmarkgen_addbenchmark(q, "null", "nullbench");

    return q;
}

void benchmarkgen_destroy(benchmarkgen _q)
{
    // free all internal benchmarks
    //if (_q->benchmarks != NULL) free(_q->benchmarks);

    // free all internal packages
    if (_q->packages != NULL) free(_q->packages);

    // free main object memory
    free(_q);
}

void benchmarkgen_parse(benchmarkgen _q,
                        char * _filename)
{
    fprintf(stderr,"benchmarkgen_parse('%s')...\n", _filename);

    // parse filename...
    benchmarkgen_parsefilename(_q, _filename);

    // parse actual file...
    benchmarkgen_parsefile(_q, _filename);
}

void benchmarkgen_print(benchmarkgen _q)
{
    unsigned int i;
    unsigned int j;
    unsigned int n;

    // count total number of benchmarks
    unsigned int num_benchmarks = 0;
    for (i=0; i<_q->num_packages; i++)
        num_benchmarks += _q->packages[i].num_benchmarks;

    // print header
    printf("// auto-generated file, do not edit\n\n");
    
    printf("#ifndef __LIQUID_BENCHMARK_INCLUDE_H__\n");
    printf("#define __LIQUID_BENCHMARK_INCLUDE_H__\n\n");

    printf("#define BENCHMARK_VERSION \"%s\"\n\n", BENCHMARK_VERSION);

    printf("// define arguments pre-processor directive\n");
    printf("#define BENCHMARK_ARGS                 \\\n");
    printf("   struct rusage * _start,             \\\n");
    printf("   struct rusage * _finish,            \\\n");
    printf("   unsigned long int * _num_iterations\n\n");

    printf("// number of packages\n");
    printf("#define NUM_PACKAGES (%u)\n\n", _q->num_packages);

    printf("// number of benchmarks\n");
    printf("#define NUM_BENCHMARKS (%u)\n\n", num_benchmarks);

    printf("// function declarations\n");
    for (i=0; i<_q->num_packages; i++) {
        struct package_s * p = &_q->packages[i];
        for (j=0; j<p->num_benchmarks; j++)
            printf("void benchmark_%s(BENCHMARK_ARGS);\n", p->benchmarks[j].name);
    }
    printf("\n");


    printf("// array of benchmarks\n");
    printf("bench_t benchmarks[NUM_BENCHMARKS] = {\n");
    n=0;
    for (i=0; i<_q->num_packages; i++) {
        struct package_s * p = &_q->packages[i];
        for (j=0; j<p->num_benchmarks; j++) {
            printf("    {%4u, &benchmark_%s,\"%s\",0,0,0,0,0}",
                n,
                p->benchmarks[j].name,
                p->benchmarks[j].name);
            if ( n < num_benchmarks-1 )
                printf(",");
            printf("\n");
            n++;
        }
    }
    printf("};\n\n");

    n=0;
    printf("// array of packages\n");
    printf("package_t packages[NUM_PACKAGES] = {\n");
    for (i=0; i<_q->num_packages; i++) {
        printf("    {%4u, %4u, %4u, \"%s\"}",
            i,
            n,
            _q->packages[i].num_benchmarks,
            _q->packages[i].name);
        if (i<_q->num_packages-1)
            printf(",");
        printf("\n");

        n += _q->packages[i].num_benchmarks;
    }
    printf("};\n\n");

    printf("#endif // __LIQUID_BENCHMARK_INCLUDE_H__\n\n");

}



//
// internal methods
//

void benchmarkgen_parsefilename(benchmarkgen _q,
                                char * _filename)
{
    fprintf(stderr,"benchmarkgen_parsefilename('%s')...\n", _filename);
    char substr[NAME_LEN];
    char _basename[NAME_LEN];
    char * subptr;
    int i0, i1;
    char pathsep = '/';
    const char tag[] = "_benchmark.h";

    // try to strip out path: find rightmost occurrence of pathsep
    //printf("%s\n", _filename);
    subptr = strrchr(_filename, pathsep);   // obtain index of last pathsep
    if (subptr == NULL) {
        //printf("path delimiter not found\n");
        i0 = 0;
    } else {
        i0 = subptr - _filename + 1;
    }
    //printf("//  i0 : %d\n", i0);

    // try to strip out tag: "_benchmark.h"
    subptr = strrchr( _filename, tag[0] );
    if (subptr == NULL) {
        fprintf(stderr,"error: benchmarkgen_parsefilename('%s'), tag not found\n", _filename);
        exit(1);
    } else {
        i1 = subptr - _filename;
        //printf("  i1 : %d\n", i1);
    }

    // ensure the last occurrence of tag is not in the path name
    if (i0 >= i1) {
        fprintf(stderr,"error: benchmarkgen_parsefilename('%s'), invalid path name\n", _filename);
        exit(1);
    }
    
    // ensure tag is valid
    //strncpy(substr,&_filename[i1],NAME_LEN);
    //printf("  comparing %s with %s\n", tag, substr);
    if (strncmp(tag,&_filename[i1],strlen(tag)) != 0 ) {
        fprintf(stderr,"error: benchmarkgen_parsefilename('%s'), invalid tag (comparison failed)\n", _filename);
        exit(1);
    } else {
        //printf("  comparison passed!\n");
    }

    // copy base name
    strncpy( substr, _filename+i0, i1-i0 );
    // add null character to end
    substr[i1-i0] = '\0';
    //printf("base: \"%s\"\n", substr);

    strncpy(_basename, substr, NAME_LEN);

    //printf("// basename : %s\n", _basename);

}

void benchmarkgen_parsefile(benchmarkgen _q,
                            char * _filename)
{
    fprintf(stderr,"benchmarkgen_parsefile('%s')...\n", _filename);
    // try to open file...
    FILE * fid = fopen(_filename,"r");
    if (!fid) {
        fprintf(stderr,"error: benchmarkgen_parsefile('%s'), could not open file for reading\n", _filename);
        exit(1);
    }

    // parse file...

    // close file
    fclose(fid);
}

void benchmarkgen_addpackage(benchmarkgen _q,
                             char * _package_name)
{
    // increase package size
    _q->num_packages++;

    // re-allocate memory for packages
    _q->packages = (struct package_s *) realloc(_q->packages,
                                                _q->num_packages*sizeof(struct package_s));

    // initialize new package
    strncpy(_q->packages[_q->num_packages-1].name,
            _package_name,
            NAME_LEN);

    // initialize number of benchmarks
    _q->packages[_q->num_packages-1].num_benchmarks = 0;
}

void benchmarkgen_addbenchmark(benchmarkgen _q,
                               char * _package_name,
                               char * _benchmark_name)
{
    // first validate that package exists
    unsigned int i;
    int package_found = 0;
    for (i=0; i<_q->num_packages; i++) {
        if ( strcmp(_q->packages[i].name, _package_name)==0 ) {
            package_found = 1;
            break;
        }
    }

    if (!package_found) {
        fprintf(stderr,"error: benchmarkgen_addbenchmark(), unknown package '%s'\n", _package_name);
        exit(1);
    }

    struct package_s * p = &_q->packages[i];

    // increase benchmark size
    p->num_benchmarks++;

    // re-allocate memory for benchmarks
    p->benchmarks = (struct benchmark_s *) realloc(p->benchmarks,
                                                   p->num_benchmarks*sizeof(struct benchmark_s));

    // initialize new benchmark
    strncpy(p->benchmarks[p->num_benchmarks-1].name,
            _benchmark_name,
            NAME_LEN);

    // associate benchmark with package
    //_q->benchmarks[_q->num_benchmarks-1].package = &_q->packages[i];
}

File: bench/benchmarkgen.py
#!/usr/bin/python
# Copyright (c) 2007, 2009 Joseph Gaeddert
# Copyright (c) 2007, 2009 Virginia Polytechnic Institute & State University
#
# This file is part of liquid.
#
# liquid is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# liquid is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with liquid.  If not, see <http://www.gnu.org/licenses/>.


# 
# Benchmark generator script
#

'''Usage: %s [OPTIONS] <input file(s)>
Generate benchmark source file.

  -v, --version         Print version and exit
  -h, --help            Print this help file and exit
  -o, --output=NAME     Write output to file NAME
'''

import re
import sys
import getopt
import glob
import string
import os.path

outputFileName = "benchmark_include.h"
included_headers = []
listed_headers = []
packages = []
benchmark_functions = []
version_number = "0.2.0"

def main():
    '''The main program'''
    files = parseCommandline()
    scanInputFiles( files )
    writeOutput()

def usage( problem = None ):
    '''Print usage info and exit'''
    if problem is None:
        print usageString()
        sys.exit(0)
    else:
        sys.stderr.write( usageString() )
        abort( str(problem) )

def usageString():
    '''Construct program usage string'''
    return __doc__ % sys.argv[0]

def abort( problem ):
    '''Print error message and exit'''
    sys.stderr.write( '\n' )
    sys.stderr.write( problem )
    sys.stderr.write( '\n\n' )
    sys.exit(2)

def ParseCommandLine():
    '''Analyze command line arguments'''
    try:
        options, patterns = getopt.getopt( sys.argv[1:], 'vho:',
                                           ['version', 'help', 'output=',] )
    except getopt.error, problem:
        usage( problem )
    setOptions( options, patterns )

def setOptions( options, patterns ):
    '''Set options specified on command line'''
    global outputFileName, listed_headers
    for o, a in options:
        if o in ('-v', '--version'):
            printVersion()
        if o in ('-h', '--help'):
            usage()
        elif o in ('-o', '--output'):
            outputFileName = a
        #elif o in ('i', '--include'):
            #if not re.match( r'^["<].*[>"]$', a ):
            #    a = ('"%s"' % a)
            #listed_headers.append( a )

    for h in patterns:
        listed_headers.append( h )

def printVersion():
    '''Print version and exit'''
    sys.stdout.write( "This is benchtest version " + version_number + "\n" )
    sys.exit(0)

class Package:
    def __init__(self, _header_file=""):
        self.header_file = _header_file

        # split base file name from path, remove extension
        self.name = os.path.splitext(os.path.basename(self.header_file))[0]
        self.benchmark_index = 0
        self.num_benchmarks = 0

    def Print(self):
        for i in range(0,self.num_benchmarks):
            benchmark_functions[self.benchmark_index+i].Print()

class BenchmarkFunction:
    def __init__(self, _function_name="", _short_name="", _file_name="", _line_number=0):
        self.function_name = _function_name
        self.short_name = _short_name
        self.file_name = _file_name
        self.line_number = _line_number

    def Print(self):
        print "  " + self.short_name + "(), " + self.file_name + ":" + str(self.line_number)

def parseHeader( filename ):
    '''Parse header file and searching for void benchmark_'''
    # add header to list
    if not filename in included_headers:
        p = Package(filename)
        p.benchmark_index = len(benchmark_functions)
    else:
        return

    if not os.path.exists(filename):
        abort("error: file does not exist: " + filename)
        return

    header_file = open(filename, 'r')
    function_key = "void benchmark_"
    this_line_number = 0
    for line in header_file.readlines():
        this_line_number += 1
        # search for "void benchmark_"
        base_index = string.find(line, function_key)
        if base_index >= 0:
            line_sub = line[len(function_key)+1:len(line)]

            # search for space or "(" or new line
            # TODO also search for tab
            end_index_newline = len(line_sub)-1
            end_index_space   = string.find(line_sub, " ")
            end_index_paren   = string.find(line_sub, "(")

            # use smallest end index greater than zero
            end_index = end_index_newline

            if end_index_space > 0:
                end_index = min(end_index,end_index_space)

            if end_index_paren > 0:
                end_index = min(end_index,end_index_paren)

            function_name = line[base_index+5:end_index+len(function_key)+1]
            short_name = line[base_index+len(function_key):end_index+len(function_key)+1]
            f = BenchmarkFunction(function_name, short_name, filename, this_line_number)
            f.Print()
            benchmark_functions.append(f)
            p.num_benchmarks += 1

    if p.num_benchmarks > 0:
        included_headers.append(filename)
        packages.append(p)

    header_file.close()

def writeOutputFile( filename = outputFileName ):
    '''Write output file'''
    f = open(outputFileName, "w")

    # write header
    f.write("// auto-generated file, do not edit\n")
    f.write("//\n")
    f.write("// created with benchtest v" + version_number + "\n\n");

    # include guard
    f.write("#ifndef __BENCHINCLUDE_H__\n");
    f.write("#define __BENCHINCLUDE_H__\n\n");
    f.write("#define BENCHMARK_VERSION " + "\"" + version_number + "\"" + "\n\n");

    #
    # benchmarks declarations
    #
    f.write("// benchmark arguments pre-processor directive\n");
    f.write("#define BENCHMARK_ARGS                 \\\n");
    f.write("   struct rusage * _start,             \\\n");
    f.write("   struct rusage * _finish,            \\\n");
    f.write("   unsigned long int * _num_iterations\n");
    f.write("\n");

    f.write("// function declarations\n");
    for i in range(len(benchmark_functions)):
        bf = benchmark_functions[i]
        # append benchmark function to array
        f.write("void " + bf.function_name + "(BENCHMARK_ARGS);\n");
    f.write("\n")

    f.write("// number of benchmarks\n");
    f.write("#define NUM_BENCHMARKS " + str(len(benchmark_functions)) + "\n\n");

    #
    # benchmarks array
    #
    f.write("// array of benchmarks\n");
    f.write("bench_t benchmarks[NUM_BENCHMARKS] = {\n")
    for i in range(len(benchmark_functions)):
        bf = benchmark_functions[i]
        # append benchmark function to array
        f.write(" "*4 + "{" + str(i) + ",&" + bf.function_name + ",\"" + bf.short_name + "\",0,0.0f,0.0f}")
        if not i == len(benchmark_functions)-1:
            f.write(",\n")
    f.write("\n};\n")

    f.write("\n")
    f.write("// number of packages\n");
    f.write("#define NUM_PACKAGES " + str(len(packages)) + "\n\n");

    f.write("// array of packages\n");
    f.write("package_t packages[NUM_PACKAGES] = {\n")
    for i in range(len(packages)):
        p = packages[i];
        # append packages to array
        f.write(" "*4 + "{" + str(i) + "," + str(p.benchmark_index) + "," + str(p.num_benchmarks) + ",\"" + p.name + "\"}")
        if not i == len(packages)-1:
            f.write(",\n")
    f.write("\n};\n")


    # include guard
    f.write("#endif // __BENCHINCLUDE_H__\n");

    f.close()
    
if __name__ == '__main__':
    if len(sys.argv) < 2:
        usage()
        sys.exit(0)

    # parse command-line arguments
    ParseCommandLine()

    # add null package/benchmark
    #p = Package("bench.c");
    p = Package("null_benchmark.h");
    p.benchmark_index = len(benchmark_functions)
    f = BenchmarkFunction("benchmark_nullbench", "nullbench", "bench.c", 0);
    f.Print();
    p.num_benchmarks += 1
    benchmark_functions.append(f);
    included_headers.append("bench.c")
    packages.append(p)

    # initialize on single file
    for h in listed_headers:
        print "parsing " + h + "..."
        parseHeader( h )

    # write output file
    if len(packages) > 0:
        writeOutputFile(outputFileName)
    else:
        print "no benchmarks could be found"


